### PR TITLE
86-bug-total-is-displayed-with-four-decimal-places

### PR DESCRIPTION
**Pull request by Segelzwerg**

### CODE REVIEW

File: invoice/models.py [hunk 1]
```diff
@@ -149,12 +149,12 @@ class InvoiceItem(Model):
     @property
     def net_total(self) -> Decimal:
         """Get the sum of the item excluding taxes."""
-        return self.price * self.quantity
+        return round(self.price * self.quantity, 2)
```
Comment (iTitus) [inline invoice/models.py:152]: does this rounding even work?
I think you need to use the decimal methods

File: invoice/models.py [hunk 2]
```diff
@@ -149,12 +149,12 @@ class InvoiceItem(Model):
...
 
     @property
     def total(self) -> Decimal:
         """Get the sum of the item including taxes."""
-        return self.net_total * (Decimal('1.0') + self.tax)
+        return round(self.net_total * (Decimal('1.0') + self.tax), 2)
```
Comment (iTitus) [inline invoice/models.py:157]: ditto

**Comment by iTitus** [comment]
Does this also fix #152 ?

File: invoice/models.py [hunk 3]
```diff
@@ -111,9 +111,9 @@ def net_total(self) -> Decimal:
         return Decimal(sum(item.net_total for item in self.items))
 
     @property
-    def total(self):
+    def total(self) -> Decimal:
         """Get the sum of total."""
-        return sum(item.total for item in self.items)
+        return Decimal(sum(item.total for item in self.items))
```
[the remark this answers is not in the document]
Comment (Segelzwerg) [reply invoice/models.py:116]: Technically, has the type hint `int`

File: invoice/models.py (hunk 1 above)
Comment (Segelzwerg) [reply invoice/models.py:152]: Actually it does. 
However, I would prefer to do it on the views/PDFs, but haven't figured out on how to do it in the PDF creation .

**Comment by Segelzwerg** [comment]
> Does this also fix #152 ?

No, but we could add it

File: invoice/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (iTitus) [reply invoice/models.py:116]: I see.
An empty sum returns `0: int`. Casting it to Decimal makes sure we stay in the same type and get no precision problems or type errors later down the line

File: invoice/models.py (hunk 1 above)
Comment (iTitus) [reply invoice/models.py:152]: I tested `round` and it works.
Explicitly rounding is bad though, we want to use a lot of precision during the calculations and only do rounding at the very end.
And we should probably use Decimal contexts

Comment (Segelzwerg) [reply invoice/models.py:152]: Yes, in theory we would never be required to round.  Unless we have fractional quantities or very small prices. So, it is always on item level and does not affected invoices totals.

File: invoice/templates/invoice/invoice_list.html [hunk 4]
```diff
@@ -22,8 +22,8 @@
                 <td>{{ invoice.date }}</td>
                 <td>{{ invoice.vendor.name }} ({{ invoice.vendor.company_name }})</td>
                 <td>{{ invoice.customer.first_name }} {{ invoice.customer.last_name }}</td>
-                <td>{{ invoice.net_total }}</td>
-                <td>{{ invoice.total }}</td>
+                <td>{{ invoice.net_total|floatformat:2 }}</td>
+                <td>{{ invoice.total|floatformat:2 }}</td>
```
Comment (iTitus) [inline invoice/templates/invoice/invoice_list.html:26]: is this floatformat working?
if it does you wouldnt need the explicit round on the property getter

Comment (Segelzwerg) [reply invoice/templates/invoice/invoice_list.html:26]: Yes it does work here, but not in the PDF generation

**Review by iTitus — APPROVED** [review]

**Pull request merged**